enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,454 @@
+   unusable. IP addresses may change at any time, making it impractical to
+   expect or require the interfaces to be manually/statically configured
+   before starting. If in single node, it also needs to read MicroShift
+   configuration to extract the service network CIDR.
+1. Have a systemd unit do the configuration.
+   This is the automatic way of doing the previous one. Introducing startup
+   order, it also needs to read MicroShift configuration and react upon IP
+   changes.
+1. Have MicroShift configure the secondary IP.
+   While MicroShift does not own the OS, all the available information to
+   properly configure the interface address (if needed) lives here. It needs
+   the service network CIDR, the API server IP, and the new configuration
+   option.
+
+### Workflow Description
+> _Disclaimer:_ Workflows described here illustrate automatic processes. This
+section is only informational about what happens when starting MicroShift.
+
+**Start MicroShift with default configuration**
+1. MicroShift service is assumed to be enabled.
+1. When starting, MicroShift will read the service CIDR from the configuration. If not set, default to `10.43.0.0/16`.
+1. Compute API service IP from the service CIDR provided. This will be the first IP from the range.
+1. MicroShift API server controller will look for loopback interface assigned IPs.
+1. If the API service IP is not assigned to the interface, do it with a `/32` netmask.
+1. Keep startup process.
+1. API server creates the endpointslices entry linked to the `kubernetes` service:
+   ```shell
+   $ KUBECONFIG=/var/lib/microshift/resources/kubeadmin/kubeconfig oc get endpointslices
+   NAME         ADDRESSTYPE   PORTS   ENDPOINTS   AGE
+   kubernetes   IPv4          6443    10.43.0.1   2d17h
+   ```
+1. Connecting to API server using external (node) IP yields external certificate.
+   ```shell
+   $ openssl s_client -connect 192.168.122.117:6443 -showcerts | openssl x509 -text -noout -in - | grep -C 1 "Alternative\|CN"
+   Can't use SSL_get_servername
+   depth=1 CN = kube-apiserver-external-signer
+   verify error:num=19:self signed certificate in certificate chain
+   verify return:1
+   depth=1 CN = kube-apiserver-external-signer
+   verify return:1
+   depth=0 CN = 192.168.122.117
+   verify return:1
+        Signature Algorithm: sha256WithRSAEncryption
+        Issuer: CN = kube-apiserver-external-signer
+        Validity
+   --
+            Not After : Jan 27 16:46:37 2024 GMT
+        Subject: CN = 192.168.122.117
+        Subject Public Key Info:
+   --
+            X509v3 Subject Alternative Name:
+                DNS:api.example.com, DNS:microshift-1, DNS:192.168.122.117, IP Address:192.168.122.117
+   ```
+1. Connecting to API server using internal IP yields internal certificate.
+   ```shell
+   $ openssl s_client -connect 10.43.0.1:6443 -showcerts | openssl x509 -text -noout -in - | grep -C 1 "Alternative\|CN"
```

Comment:
If I've understood correctly, `10.43.0.1` is the clusterIP for kubernetes.default.svc. This IP and the matching certificate match my expectations.

Reply:
Yes. In the default case the service network CIDR is `10.43.0.0/16` and the API server IP in that range is `10.43.0.1`. Both the clusterIP and the endpoint slice match it.